File: 0207. Course Schedule.cpp
class Solution 
{
public:
    vector<vector<int>> g;
    vector<int> c;
    
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) 
    {
        c = vector<int>(numCourses, 0);
        g = vector<vector<int>>(numCourses, vector<int>());
        
        for (int i = 0; i < prerequisites.size(); ++i)
        {
            g[prerequisites[i][1]].push_back(prerequisites[i][0]);
        }
        
        for (int i = 0; i < numCourses; ++i)
        {
            if (!c[i])
            {
                if (!dfs(i))
                {
                    return false;
                }
            }
        }
        
        return true;
    }
    
    bool dfs(int v)
    {
        c[v] = 1;
        
        for (int i = 0; i < g[v].size(); ++i)
        {
            int u = g[v][i];

            if (!c[u])
            {
                if (!dfs(u))
                {
                    return false;
                }
            }
            else if (c[u] == 1)
            {
                return false;
            }
        }
        
        c[v] = 2;
        
        return true;
    }
};

File: 0207. Course Schedule.cs
public class Solution
{
    public bool CanFinish(int numCourses, int[][] prerequisites)
    {
        var states = new int[numCourses]; // 0 - not visited, 1 - in, 2 - out
        var graph = Enumerable.Range(0, numCourses).Select(x => new List<int>()).ToList();

        foreach (var prerequisite in prerequisites)
        {
            graph[prerequisite[0]].Add(prerequisite[1]);
        }

        for (int course = 0; course < numCourses; ++course)
        {
            if (states[course] == 0)
            {
                if (!TrySort(course, graph, states))
                {
                    return false;
                }
            }
        }

        return true;
    }

    private bool TrySort(
        int v,
        List<List<int>> graph,
        int[] states)
    {
        if (states[v] == 0)
        {
            states[v] = 1;

            foreach (var u in graph[v])
            {
                if (!TrySort(u, graph, states))
                {
                    return false;
                }
            }

            states[v] = 2;
        }
        else if (states[v] == 1)
        {
            return false;
        }
        
        return true;
    }
}